src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -166,10 +235,229 @@ private boolean validateNucleusConfig(CompletableFuture<DeploymentResult> totall
+                return false;
+            } catch (CertificateException | IOException |
+                    KeyStoreException | NoSuchAlgorithmException
+                     | TLSAuthException | InvalidEnvironmentStageException e) {
+                logger.atError().cause(e).log("Unable to create GGv2 data client");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+                return false;
+            } catch (InterruptedException e) {
+                logger.atError().cause(e).log("Validation interrupted unexpectedly");
+                totallyCompleteFuture
+                        .complete(new DeploymentResult(DeploymentResult.DeploymentStatus.FAILED_NO_STATE_CHANGE, e));
+                return false;
+            }
+            connection.close();
         }
         return true;
     }
 
+    private void checkHttpConnection(GreengrassV2DataClient greengrassV2DataClient, long timeout) throws InterruptedException, ComponentConfigurationValidationException {
+        ListThingGroupsForCoreDeviceRequest request = ListThingGroupsForCoreDeviceRequest.builder()
+                .coreDeviceThingName(Coerce.toString(deviceConfiguration.getThingName()))
+                .build();
+        long tryUntil = System.currentTimeMillis() + timeout;
+        while(System.currentTimeMillis() < tryUntil) {
+            try {
+                greengrassV2DataClient.listThingGroupsForCoreDevice(request);
+                return;
+            } catch (GreengrassV2DataException | SdkClientException e) {
+                Thread.sleep(TimeUnit.SECONDS.toMillis(5));
+            }
+        }
+        throw new ComponentConfigurationValidationException("HTTP client failed to connect to IoT Core using new configurations", DeploymentErrorCode.FAILED_TO_RECONNECT);
+    }
+
+    private void checkMqttConnection(MqttClientConnection connection, long timeout) throws InterruptedException, ComponentConfigurationValidationException {
+        long tryUntil = System.currentTimeMillis() + timeout;
+        while(System.currentTimeMillis() < tryUntil) {
+            try {
+                connection.connect().get();
+                return;
+            } catch (MqttException | ExecutionException e) {
+                Thread.sleep(TimeUnit.SECONDS.toMillis(2));
+            }
+        }
+        throw new ComponentConfigurationValidationException("MQTT client failed to connect to IoT Core using new configurations", DeploymentErrorCode.FAILED_TO_RECONNECT);
+    }
+
+    private MqttClientConnection createMqttConnection(Map<String, Object> nucleusConfig) {
+        AwsIotMqttConnectionBuilder builder;
+        try {
+            builder = securityService.getDeviceIdentityMqttConnectionBuilder();
+        } catch (MqttConnectionProviderException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While re-throwing the caught exception with modifications , information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)

Suggested remediation:
Consider passing the exception object either as  &#34;rootCause&#34; or as an inner exception parameter, to the constructor of the new exception before throwing new exception. This will help keep stack trace.
```
@@ -315,3 +315,3 @@
         } catch (MqttConnectionProviderException e) {
-            throw new MqttException(e.getMessage());
+            throw new MqttException(e.getMessage(), e);
         }

```